### ISSUE TITLE
Install any node version from nvm + global dependencies?

### ISSUE THREAD

**Issue by hanoii** [drud/ddev#4250]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

Nope

### Describe your solution

I got curious into how node was installed and found it on the code. Why did you take the approach of

```
RUN (apt-get remove -y nodejs || true) && (apt purge nodejs || true)
RUN curl -sSL --fail https://deb.nodesource.com/setup_14.x >/tmp/setup_node.sh ||  curl -sSL --fail https://deb.nodesource.com/setup_14.sh >/tmp/setup_node.sh
RUN bash /tmp/setup_node.sh && apt-get install nodejs && npm config set unsafe-perm true && npm install --global gulp-cli yarn
```

Why not simply do it through nvm and allow ANY version on `nodejs_version`? Also I wonder if something like this: https://docs.platform.sh/languages/nodejs.html#2-specify-any-global-dependencies could be added to config which could work for node/php and potentially others in the future?

**Comment by rfay** [comment]
nodejs_version is done with the official packages. It predated the nvm support. So I guess it's just historical. 

Note that nodejs when used from packages is done at Dockerfile time (so no overhead on second start), and nvm is done at start time (so it always has some overhead). 

I don't know how nvm could set up at image build time and still be changeable by the user, which I'd assume they'd expect.

**Comment by hanoii** [comment]
Different thoughts:

1. I can't think of a reason one would want to switch node version within the same project other than on build time. It is, if anything, very very rare. But let's assume you want to.
2. If you were to change versions, I assume you can sudo. I wonder if you install several versions from Dockerfile (like: `nvm install 16`, `nvm install 8`) you can then `nvm use` without sudo. Or you can just give user ownnership to nvm.
3. I believe you can use NVM in Dockerfile AND nvm in the user space. Obviously a bit of tweaking into how and where you install nvm and how you use the cache needs work, but I'd think it very similar to https://docs.platform.sh/languages/nodejs/node-version.html#use-nvm.

**Comment by hanoii** [comment]
Not sure about #3, but I am doing something very similar to 
https://github.com/hanoii/ddev-nvm-issue-4250/blob/main/.ddev/web-build/Dockerfile.nvm And it works, I can even install another version because of the chmod

This doesn't use the cache, but I challenge that as well, do we need the cache at all if we are installing necessary version on the image? I really really don't think you need to `nvm install` in the lifetime of a project, only initially.

**Comment by hanoii** [comment]
to add to the cache question, we are caching nvm per project, so not sure if the cache there is worth it

**Comment by rfay** [comment]
I think if we can keep backward compatibility, a PR would be fine. I do think people expect to be able to use nvm and `ddev nvm` on an existing project.

**Comment by kevin-appelt** [comment]
Is there a status update here?
It would be great if we could specify a more specific Node version. The corresponding version is also reused by us in other contexts (f.e. in our GitLab Build-Chain) and sometimes there is unexpected behaviour because the node or npm version in DDEV is slightly different than expected due to an update - even if the major version is still identical.
If we could specify the exact version as with the Composer version, both Node version and NPM version are fixed and we can use the same version everywhere.
I have tried to solve the problem with the currently available NVM integration but have not found a practical solution that may not be at least somewhat confusing (e.g. because the specified nodejs_version does not match the nvm version).

**Comment by rfay** [comment]
Hi @kevin-appelt - Why don't you just use `ddev nvm install <version>` ? See https://ddev.readthedocs.io/en/latest/users/basics/cli-usage/#nodejs-npm-nvm-and-yarn

**Comment by kevin-appelt** [comment]
Hi @rfay 
My issue is that the change has to be done not only for me, but also for all my colleagues. In addition, the version is not identical in all projects. So the best would be to have the version in the DDEV configuration and in the Git repository.
There is already the `nodejs_version`, so it seems to be a good idea to use it.
I tried a post_start hook that uses `ddev nvm install <version>`, but it seems to work only at the first start of the project. On each subsequent start I get an error message that the version is already installed and when I then run `node --version` in the web container I seem to get the version defined by `nodejs_version`. 
All in all it feels more like a workaround than a real solution (for us), because with `nodejs_version` a basically working configuration already exists - only limited to the major version.
A change of the version at runtime is (again: for us) not necessary either. The container runs with the specified version until this version gets updated.

**Comment by rfay** [comment]
If you want an explicit patch version of node,  you can put it in a post-start hook and check in your .ddev/config.yaml as is normally done.
```yaml
hooks:
  post-start:
  - exec: "nvm install <version> && nvm use <version>"
```

If you really want an explicit version of the nodejs package, you can do it with a .ddev/web-build/Dockerfile.nodejs, which would do it at the package level.

**Comment by hanoii** [comment]
This is my `Dockerfile.node` on a project I need node 8
```
ENV NVM_DIR=/usr/local/nvm
ENV NODE_DEFAULT_VERSION=8.17

RUN (apt-get remove -y nodejs || true) && (apt purge nodejs || true)

RUN curl -sL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.1/install.sh -o install_nvm.sh
RUN mkdir -p $NVM_DIR && bash install_nvm.sh
RUN echo "source $NVM_DIR/nvm.sh" >>/etc/profile
RUN bash -ic "nvm install $NODE_DEFAULT_VERSION && nvm use $NODE_DEFAULT_VERSION"
RUN chmod -R ugo+w $NVM_DIR
```

**Comment by jamesmacwhite** [comment]
> ```yaml
> hooks:
>   post-start:
>   - exec: "nvm install <version> && nvm use <version>"
> ```

When trying to use nvm from a post-start hook command I always get:

```
bash: line 1: nvm: command not found
```

If I do `ddev nvm ...` it works and likewise if I SSH into the web container and run `nvm` it's also OK, is there anything special that needs to be done beforehand to make this command work?

**Comment by rfay** [comment]
So nvm is unfortunately not a regular command that can be executed, it's a bash alias. But please do open new issues instead of commenting on unrelated issues. You can work around your issue like this:
```
hooks:
  post-start:
  - exec-host: ddev nvm install 16
```

**Comment by jamesmacwhite** [comment]
@rfay My apologies, as this issue was still open and tangibly related to the subject, I didn't want to create a whole other issue for fear of further noise ironically.

Thank you for the workaround.

**Comment by timnolte** [comment]
> So nvm is unfortunately not a regular command that can be executed, it's a bash alias. But please do open new issues instead of commenting on unrelated issues. You can work around your issue like this:
> 
> ```
> hooks:
>   post-start:
>   - exec-host: ddev nvm install 16
> ```

@rfay so this doesn't really work. I need to use Node 10, don't get me started on the EOL of it, and simply trying to use NVM and even having it set 10 as the default doesn't work when later I attempt to have a command that runs in the `web` service do `npm ci` for example, it's still running Node 14(what's defined in the `config.yml`). I need a solution that isn't going to require a huge number of manual steps every time we need to install packages are run builds of the code.

**Comment by rfay** [comment]
Could you please try latest HEAD version of DDEV? A significant fix was made (fixing a regression in v1.21.4). You can get latest at https://nightly.link/drud/ddev/workflows/master-build/master

See 
* #4620

I confirm that this works with HEAD:

```yaml
hooks:
  post-start:
    - exec: set +u && nvm install 10 && nvm alias default 10
    - exec: node --version
```

**Comment by timnolte** [comment]
@rfay so the `HEAD` version is working however there seems to be some other issue in `HEAD`. I'm not sure if this has been reported another issue but I'm getting tons of duplicate output during a `ddev start` & `ddev stop`

```shell
 Network ddev-<project>_default  Creating 
 Network ddev-<project>_default  Created 
 Container ddev-<project>-db  Creating 
 Container ddev-<project>-redis  Creating 
 Container ddev-<project>-web  Creating 
 Container ddev-<project>-dba  Creating 
 Container ddev-<project>-web  Created 
 Container ddev-<project>-db  Created 
 Container ddev-<project>-dba  Created 
 Container ddev-<project>-redis  Created 
 Container ddev-<project>-redis  Starting 
 Container ddev-<project>-web  Starting 
 Container ddev-<project>-dba  Starting 
 Container ddev-<project>-db  Starting 
 Container ddev-<project>-web  Started 
 Container ddev-<project>-dba  Started 
 Container ddev-<project>-db  Started 
 Container ddev-<project>-redis  Started 
 Container ddev-router  Creating 
 Container ddev-router  Created 
 Container ddev-router  Starting 
 Container ddev-router  Started 
```

```shell
 Container ddev-<project>-web  Stopping 
 Container ddev-<project>-redis  Stopping 
 Container ddev-<project>-db  Stopping 
 Container ddev-<project>-dba  Stopping 
 Container ddev-<project>-redis  Stopped 
 Container ddev-<project>-web  Stopped 
 Container ddev-<project>-dba  Stopped 
 Container ddev-<project>-db  Stopped 
 Container ddev-<project>-redis  Stopping 
 Container ddev-<project>-redis  Stopping 
 Container ddev-<project>-web  Stopping 
 Container ddev-<project>-db  Stopping 
 Container ddev-<project>-web  Stopping 
 Container ddev-<project>-db  Stopping 
 Container ddev-<project>-dba  Stopping 
 Container ddev-<project>-dba  Stopping 
 Container ddev-<project>-redis  Stopped 
 Container ddev-<project>-redis  Removing 
 Container ddev-<project>-db  Stopped 
 Container ddev-<project>-dba  Stopped 
 Container ddev-<project>-db  Removing 
 Container ddev-<project>-dba  Removing 
 Container ddev-<project>-web  Stopped 
 Container ddev-<project>-web  Removing 
 Container ddev-<project>-redis  Removed 
 Container ddev-<project>-db  Removed 
 Container ddev-<project>-dba  Removed 
 Container ddev-<project>-web  Removed 
 Network ddev-<project>_default  Removing 
 Network ddev-<project>_default  Removed 
```

**Comment by rfay** [comment]
I think there's too much of that, please open an issue, thanks.
